lti_consumer/lti_xblock.py

```diff
@@ -1100,172 +1099,70 @@ def lti_launch_handler(self, request, suffix=''):  # pylint: disable=unused-argu
-            # different parameters, set by instructors when running the DL configuration flow.
-            elif self.lti_advantage_deep_linking_enabled and 'deep_linking_content_launch' in lti_message_hint:
-                # Retrieve Deep Linking parameters using lti_message_hint parameter.
-                deep_linking_id = lti_message_hint.split(':')[1]
-                from lti_consumer.api import get_deep_linking_data  # pylint: disable=import-outside-toplevel
-                dl_params = get_deep_linking_data(deep_linking_id, block=self)
-
-                # Modify LTI launch and set ltiResourceLink parameters
-                lti_consumer.set_dl_content_launch_parameters(
-                    url=dl_params.get('url'),
-                    custom=dl_params.get('custom')
-                )
-
-            # Update context with LTI launch parameters
-            context.update({
-                "preflight_response": preflight_response,
-                "launch_request": lti_consumer.generate_launch_request(
-                    resource_link=str(self.location),  # pylint: disable=no-member
-                    preflight_response=preflight_response
-                )
-            })
-
-            template = loader.render_mako_template('/templates/html/lti_1p3_launch.html', context)
-            return Response(template, content_type='text/html')
-        except Lti1p3Exception as exc:
-            log.warning(
-                "Error preparing LTI 1.3 launch for block %r: %s",
-                str(self.location),  # pylint: disable=no-member
-                exc,
-            )
-            template = loader.render_django_template('/templates/html/lti_1p3_launch_error.html', context)
-            return Response(template, status=400, content_type='text/html')
-        except AssertionError as exc:
-            log.warning(
-                "Permission on LTI block %r denied for user %r: %s",
-                str(self.location),  # pylint: disable=no-member
-                self.external_user_id,
-                exc,
-            )
-            template = loader.render_django_template('/templates/html/lti_1p3_permission_error.html', context)
-            return Response(template, status=403, content_type='text/html')
+        # Backward compatibility - copy the config to the DB
+        self._save_lti_1p3_config_to_db()
+
+        from .plugin.views import launch_gate_endpoint  # pylint: disable=import-outside-toplevel
+        request.user = self.runtime.service(self, 'user').get_current_user()
+        return launch_gate_endpoint(request, suffix)
+
+    def _save_lti_1p3_config_to_db(self):
+        """
+        Save LTI 1.3 config to the database.
+        """
+        from .api import _get_lti_config  # pylint: disable=import-outside-toplevel
+        conf = _get_lti_config(block=self)
+        # Transfer the config to DB so that the django view can initialize the LTI1P3Consumer
```

Comment:
I have improved the [DB update logic](https://github.com/openedx/xblock-lti-consumer/pull/254/commits/c7a85450202c265ea41d12b1574444a86dc658d7#diff-6e408862179502a0ac809fab914e028b18311a17a3b0874469bd6416bc4a449fR114). It is done in the following scenarios:

1. Every time the [XBlock is updated](https://github.com/openedx/xblock-lti-consumer/pull/254/commits/c7a85450202c265ea41d12b1574444a86dc658d7#diff-cc618c2e63b8c2d293d245f9799381d9321e40fe2d05e9a4b2f80bb91995157dR1052) from Studio using the Edit button
2. In order to maintain backward compatibility with existing integrations, the `get_or_update_lti_config` will be called in every request passing through the XBlock Handlers (preflight_launch_handler or access_token_handler), but the update will only in the first call.

If someone manually changes values in `lti_config` attribute via the admin, and it has corresponding values in the XBlock, then the XBlock values will be writer OVER the MANUALly edited values. Is this something of a concern? I am assuming the maintainers are more likely to edit the XBlock in Studio to change LTI 1.3 configurations, than to tinker with values in the admin :thinking:

Reply:
@tecoholic,

> If someone manually changes values in lti_config attribute via the admin, and it has corresponding values in the XBlock, then the XBlock values will be writer OVER the MANUALly edited values. Is this something of a concern? I am assuming the maintainers are more likely to edit the XBlock in Studio to change LTI 1.3 configurations, than to tinker with values in the admin thinking

Good point. Doesn't this make DB configuration unusable? If somebody changes this through the admin interface, then even updating this XBlock's Display Name is going to overwrite all changes, no? What if we added something like this, e.g. as a `post_save` receiver?
```python
@receiver(post_save, sender=LtiConfiguration)
def update_xblock_config(sender, instance, **kwargs):
    block = instance.block
    # Update configs.
    block.save()
```

Reply:
@Agrendalath I have implemented the hook in [3976c2b](https://github.com/openedx/xblock-lti-consumer/pull/254/commits/3976c2b1102878009a9b399b96ff89ad2954b40f) using the `pre_save` signal instead of the `post_save`, so that the side-effects can easily be managed.

Initially I tried with `post_save` and then patched all the test cases where this was causing a side-effect. But it was still causing a lot of test-case failures. The only way I could find to totally isolate the situation was to determine if the XBlock really needed updating, by comparing the previous value with the current value. But it's not possible from the `post_save` as the DB is already updated.

So I have used `pre_save` which allows us to fetch the XBlock and update it only when specific fields are changed. Right now I have limited it to the values stored in `lti_config` attribute. But after #260 is merged, I think this will change.

Reply:
This part is now removed.